pylint/checkers/variables.py

```diff
@@ -665,6 +673,146 @@ def get_next_to_consume(self, node: nodes.Name) -> list[nodes.NodeNG] | None:
 
         return found_nodes
 
+    @staticmethod
+    def _exhaustively_define_name_raise_or_return(
+        name: str, node: nodes.NodeNG
+    ) -> bool:
+        """Return True if there is a collectively exhaustive set of paths under
+        this `if_node` that define `name`, raise, or return.
+        """
+        # Handle try and with
+        if isinstance(node, (nodes.TryExcept, nodes.TryFinally)):
+            # Allow either a path through try/else/finally OR a path through ALL except handlers
+            return (
+                NamesConsumer._defines_name_raises_or_returns_recursive(name, node)
+                or isinstance(node, nodes.TryExcept)
+                and all(
+                    NamesConsumer._defines_name_raises_or_returns_recursive(
+                        name, handler
+                    )
+                    for handler in node.handlers
+                )
+            )
+        if isinstance(node, nodes.With):
+            return NamesConsumer._defines_name_raises_or_returns_recursive(name, node)
+
+        if not isinstance(node, nodes.If):
+            return False
+
+        # Be permissive if there is a break
+        if any(node.nodes_of_class(nodes.Break)):
+            return True
+
+        # If there is no else, then there is no collectively exhaustive set of paths
+        if not node.orelse:
+            return False
+
+        # Is there an assignment in this node itself, e.g. in named expression?
+        if NamesConsumer._defines_name_raises_or_returns(name, node):
+            return True
+
+        return NamesConsumer._branch_handles_name(
+            name, node.body
+        ) and NamesConsumer._branch_handles_name(name, node.orelse)
+
+    @staticmethod
+    def _branch_handles_name(name: str, body: Iterable[nodes.NodeNG]) -> bool:
+        return any(
+            NamesConsumer._defines_name_raises_or_returns(name, if_body_stmt)
+            for if_body_stmt in body
+        ) or (
+            any(
+                NamesConsumer._exhaustively_define_name_raise_or_return(name, nested)
+                for nested in body
+                if isinstance(
+                    nested,
+                    (nodes.If, nodes.TryExcept, nodes.TryFinally, nodes.With),
+                )
+            )
+        )
```

Comment:
```suggestion
        return any(
            NamesConsumer._defines_name_raises_or_returns(name, if_body_stmt)
            or isinstance(
                if_body_stmt,
                (nodes.If, nodes.TryExcept, nodes.TryFinally, nodes.With),
            )
            and NamesConsumer._exhaustively_define_name_raise_or_return(
                name, if_body_stmt
            )
            for if_body_stmt in body
        )
```
Can't we do that and loop only once instead of twice ?